Sources/Valkey/Subscriptions/SubscriptionConnectionStateMachine.swift

```diff
@@ -0,0 +1,258 @@
+            case .cancel(let cont):
+                cont.resume(throwing: CancellationError())
+            case .release(let continuation):
+                continuation.resume()
+                self.logger.trace("Released connection for subscriptions")
+            case .doNothing:
+                break
+            }
+        }
+    }
+}
+
+/// StateMachine for acquiring Subscription Connection.
+@usableFromInline
+struct SubscriptionConnectionStateMachine<Value, Request, ReleaseRequest>: ~Copyable {
+    enum State: ~Copyable {
+        /// We have no connection
+        case uninitialized
+        /// We are acquiring a connection
+        case acquiring([Int: Request])
+        /// We have a connection
+        case acquired(Value, Set<Int>, ReleaseRequest)
+    }
+    var state: State
+
+    init() {
+        self.state = .uninitialized
+    }
+
+    init(state: consuming State) {
+        self.state = state
+    }
+
+    enum GetAction {
+        case startAcquire
+        case doNothing
+        case completeRequest(Value)
+    }
+
+    mutating func get(id: Int, request: Request) -> GetAction {
+        switch consume self.state {
+        case .uninitialized:
+            self = .acquiring([id: request])
+            return .startAcquire
+        case .acquiring(var map):
+            map[id] = request
+            self = .acquiring(map)
+            return .doNothing
+        case .acquired(let connection, var ids, let releaseRequest):
+            ids.insert(id)
+            self = .acquired(connection, ids, releaseRequest)
+            return .completeRequest(connection)
+        }
+    }
+
+    enum CancelAction {
+        case cancel(Request)
+        case release(ReleaseRequest)
+        case doNothing
+    }
+
+    mutating func cancel(id: Int) -> CancelAction {
+        switch consume self.state {
+        case .uninitialized:
+            self = .uninitialized
+            return .doNothing
+        case .acquiring(var map):
+            guard let continuation = map.removeValue(forKey: id) else {
+                self = .acquiring(map)
+                return .doNothing
+            }
+            if map.isEmpty {
+                self = .uninitialized
+            } else {
+                self = .acquiring(map)
+            }
+            return .cancel(continuation)
+        case .acquired(let connection, var ids, let releaseRequest):
+            ids.remove(id)
+            if ids.isEmpty {
+                self = .uninitialized
+                return .release(releaseRequest)
+            } else {
+                self = .acquired(connection, ids, releaseRequest)
+                return .doNothing
+            }
+        }
+    }
+
+    enum AcquiredAction {
+        case yield([Request])
+    }
+
+    mutating func acquired(_ value: Value, releaseRequest: ReleaseRequest) -> AcquiredAction {
+        switch consume self.state {
+        case .uninitialized:
+            fatalError()
```

Comment:
can we print the state in all of those? this will make future debugging much simpler.